tests/dependency-tests/root.rs

```diff
@@ -0,0 +1,294 @@
+use std::process::Command;
+
+#[derive(Debug)]
+enum Search<'a> {
+    Positive(&'a str),
+    Negative(&'a str),
+}
+
+#[derive(Debug)]
+struct Requirement<'a> {
+    human_readable_name: &'a str,
+    target: &'a str,
+    packages: &'a [&'a str],
+    features: &'a [&'a str],
+    default_features: bool,
+    search_terms: &'a [Search<'a>],
+}
+
+fn check_feature_dependency(requirement: Requirement) {
+    println!("Checking: {}", requirement.human_readable_name);
+
+    let mut args = Vec::new();
+    args.extend(["tree", "--target", requirement.target]);
+
+    for package in requirement.packages {
+        args.push("--package");
+        args.push(package);
+    }
+
+    if !requirement.default_features {
+        args.push("--no-default-features");
+    }
+
+    let features = requirement.features.join(",");
+    if !requirement.features.is_empty() {
+        args.push("--features");
+        args.push(&features);
+    }
+
+    println!("$ cargo {}", args.join(" "));
+
+    let output = Command::new("cargo")
+        .args(&args)
+        .output()
+        .expect("Failed to run cargo tree")
+        .stdout;
+    let output = String::from_utf8(output).expect("Output is not valid UTF-8");
+
+    let mut any_failed = false;
+    println!("{output}");
+
+    for (i, search_term) in requirement.search_terms.iter().enumerate() {
+        // Add a space and after to make sure we're getting a full match
+        let found = match search_term {
+            Search::Positive(search_term) => output.contains(&format!(" {search_term} ")),
+            Search::Negative(search_term) => !output.contains(&format!(" {search_term} ")),
+        };
+
+        if found {
+            println!(
+                "✅ Passed! ({} of {})",
+                i + 1,
+                requirement.search_terms.len()
+            );
+        } else {
+            println!(
+                "❌ Failed! ({} of {})",
+                i + 1,
+                requirement.search_terms.len()
+            );
+            any_failed = true;
+        }
+    }
+
+    assert!(!any_failed);
+}
+
+fn get_all_wgpu_features() -> Vec<String> {
+    let metadata = cargo_metadata::MetadataCommand::new()
+        .no_deps()
+        .exec()
+        .unwrap();
+
+    metadata
+        .packages
+        .iter()
+        .find_map(|p| {
+            if p.name == "wgpu" {
+                Some(p.features.keys().cloned().collect())
+            } else {
+                None
+            }
+        })
```

Comment:
nitpick: Sorta weird, this is a place where `filter`/`map` is _not_ actually worse than `filter_map`. 😅

Reply:
Fair, willfix